torch/distributed/fsdp/_common_utils.py

```diff
@@ -87,6 +89,21 @@ def _get_module_fsdp_state_if_fully_sharded_module(
     return None
 
 
+# Get all the sharded submodule names under the 'root_module' that is usually passed to 'fully_shard()'
```

Comment:
I want to clarify this comment (and similarly for the other helper).

Is the expectation that the user only passes `module` as the exact module passed to `fully_shard()`? Is this a hard requirement, or is this just the expected usage?

---

My concern with the current implementation is that:
- `_fully_sharded_submodule_names` can be accessed by any submodule managed by `fully_shard()`.
- `_fully_sharded_submodule_names` is not shared across different manual `fully_shard()` calls.

What this means is that a user should be able to call `_get_fully_sharded_submodule_names()` on any submodule managed by `fully_shard()` without erroring but will only see the expected full list of submodule names if using auto wrapping, not manual wrapping.

The reason is that (maybe non-intuitively), only for auto wrapping do submodules managed by `fully_shard()` share the same `_FSDPState` object. The design to share the same `_FSDPState` object is a helpful step for implementing any non-recursive wrapping design, but at the moment, we just have this design difference between wrapper FSDP and `fully_shard()`.

(If you add a unit test that uses manual `fully_shard()` "wrapping", I think the helper will not return the list of names you expect.)

Reply:
what you are saying is correct, I did have to call the helper twice if I call fully_shard() twice one submodule first and then root module (not manual wrapping in the mode codes, but the two calls are composed/nested).

I think it probably makes more sense to get "_all_fsdp_states" to get all the children's sharding info and managed parameters in this case. as usually I just want to get this debug info in one call